website/notifications/events/model.py

```diff
@@ -0,0 +1,352 @@
+
+from furl import furl
+from datetime import datetime
+from six import add_metaclass
+
+from website.notifications import emails
+from website.notifications.constants import NOTIFICATION_TYPES
+from website.models import Node
+from website.notifications.events import utils as event_utils
+from website.notifications import utils
+
+
+class _EventMeta(type):
+    registry = {}
+
+    def __init__(cls, name, bases, attrs):
+        if name != 'Event':
+            event_id = name.lower()
+            cls.registry[event_id] = cls
+
+        super(_EventMeta, cls).__init__(name, bases, attrs)
+
+
+@add_metaclass(_EventMeta)
+class Event(object):
+    """
+    Base notification class for building notification events and messages.
+    - abstract methods set methods that must be defined by subclasses.
+
+    To use this interface you must use the class as a Super (inherited).
+     - Implement property methods in subclasses
+     - All subclasses must be in this file for the meta class to list them
+     - Name the subclasses you will be calling as such:
+      - event (the type of event from _SUBSCRIPTIONS_AVAILABLE or specific cases)
+      - class
+      example: event = file_added, class = FileAdded
+    """
+
+    def __init__(self, user, node, action):
+        self.user = user
+        self.gravatar_url = user.gravatar_url
+        self.node = node
+        self.action = action
+        self.timestamp = datetime.utcnow()
+
+    @classmethod
+    def parse_event(cls, user, node, event, **kwargs):
+        """If there is a class that matches the event then it returns that instance"""
+        kind = ''.join(event.split('_'))
+        if kind in cls.registry:
+            return cls.registry[kind](user, node, event, **kwargs)
+        raise TypeError
+
+    def perform(self):
+        """Calls emails.notify to notify users of an action"""
+        emails.notify(
+            event=self.event,
+            user=self.user,
+            node=self.node,
+            timestamp=self.timestamp,
+            message=self.html_message,
+            gravatar_url=self.gravatar_url,
+            url=self.url.url
+        )
+
+    @property
+    def text_message(self):
+        raise NotImplementedError
+
+    @property
+    def html_message(self):
+        raise NotImplementedError
+
+    @property
+    def url(self):
+        raise NotImplementedError
+
+    @property
+    def event(self):
+        """String
+
+        Examples:
+            <waterbutler id>_file_updated"""
+        raise NotImplementedError
+
+
+class FileEvent(Event):
+    """File event base class, should not be called directly"""
+
+    def __init__(self, user, node, event, payload=None):
+        super(FileEvent, self).__init__(user, node, event)
+        self.payload = payload
+        self._url = None
+
+    @property
+    def html_message(self):
+        """Most basic html message"""
+        f_type, action = self.action.split('_')
```

Comment:
More magic here; try to avoid dependence on naming conventions.

If you decide that you absolutely, positively cannot avoid this, then make sure to clearly document all these contracts in the class's docstring.